<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">

<!--  Version: Multiflex-5.4 / About                     -->
<!--  Type:    Design with sidebar                          -->
<!--  Date:    March 13, 2008                               -->
<!--  Design:  www.1234.info                                -->
<!--  License: Fully open source without restrictions.      -->
<!--           Please keep footer credits with the words    -->
<!--           "Design by 1234.info". Thank you!            -->

<head>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-25486066-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
  <meta http-equiv="content-type" content="text/html; charset=utf-8" />
  <meta http-equiv="cache-control" content="no-cache" />
  <meta http-equiv="expires" content="3600" />
  <meta name="revisit-after" content="2 days" />
  <meta name="robots" content="index,follow" />
  <meta name="publisher" content="Your publisher infos here" />
  <meta name="copyright" content="Copyright (c) 2011 Adam Roberts" />
  <meta name="author" content="Designed by www.1234.info / Modified: Adam Roberts" />
  <meta name="distribution" content="global" />
  <meta name="image" content="http://bio.math.berkeley.edu/eXpress/img/logo.png" />
  <meta name="description" content="eXpress is a general quantification tool for target DNA/RNA sequences. While its primary use currently is RNA-Seq it has the potential for applications in many other areas including as allele-specific expression and metgenomics. What makes eXpress different is that it is an online (or streaming) algorithm, meaning it only makes one pass through the data. This allows it to be very light-weight and efficient using a constant amount of memory and time linear in the number of sequenced fragments being processed. Furthermore, it accepts piped SAM/BAM input, allowing users to avoid storing extremely large alignment files. eXpress models fragment biases, fragment lengths, and errors, allowing it to also be one of the most accurate quantification methods available." />
  <meta name="keywords" content="RNA-Seq, Genomics, transcript,quantification" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_reset.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_grid.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_content.css" />
  <link rel="icon" type="image/x-icon" href="./img/favicon.ico" />
  <title>eXpress &bull; FAQ</title>
   <script language="Javascript">
        logo1=new Image
        logo1.src="img/logo.png"
        logo2=new Image
        logo2.src="img/logo_yellow.png"
 </script>
  <title>eXpress &bull; FAQ</title>
</head>

<!-- Global IE fix to avoid layout crash when single word size wider than column width -->
<!-- Following line MUST remain as a comment to have the proper effect -->
<!--[if IE]><style type="text/css"> body {word-wrap: break-word;}</style><![endif]-->



<body>
  <!-- CONTAINER FOR ENTIRE PAGE -->
  <div class="container">

    <!-- A. HEADER -->         
    <div class="corner-page-top"></div>        
    <div class="header">
      <div class="header-top">
        
        <!-- A.1 SITENAME -->    
        <div class="sitelogo">
        <ul>
            <li><a href="#" onMouseOver="document.logo.src=logo2.src" onMouseOut="document.logo.src=logo1.src"><img name="logo" src="img/logo.png"/></a></li>
        </ul>
        </div>
        <div class="sitename">
          <h1><a href="#">eXpress</a></h1> 
          &nbsp;
          <h2><i>Streaming</i> quantification for high-throughput sequencing</h2>
        </div>
    
        <!-- A.2 BUTTON NAVIGATION -->
        <div class="navbutton">
          <ul>
            <li><a href="http://www.berkeley.edu"><img src="img/berkeley_seal.gif"/></a></li>
          </ul>
        </div>
      </div>

      <!-- A.4 BREADCRUMB and SEARCHFORM -->
      <div class="header-bottom">


        <!-- Search form -->    
<div class="searchform" id="cse-search-form" style="padding-top:4px; width:30%;">Loading</div>
<script src="http://www.google.com/jsapi" type="text/javascript"></script>
<script type="text/javascript"> 
  google.load('search', '1', {language : 'en'});
  google.setOnLoadCallback(function() {
    var customSearchControl = new google.search.CustomSearchControl('017113487431020499685:jdftct1emrw');
    customSearchControl.setResultSetSize(google.search.Search.FILTERED_CSE_RESULTSET);
    var options = new google.search.DrawOptions();
    options.enableSearchboxOnly("http://www.google.com/cse?cx=017113487431020499685:jdftct1emrw", null, true);
    customSearchControl.draw('cse-search-form', options);
  }, true);
</script>
<link rel="stylesheet" href="css/googlesearch.css" type="text/css" />

   </div>
    </div>      
    <div class="corner-page-bottom"></div>    
    
    <!-- B. NAVIGATION BAR -->
    <div class="corner-page-top"></div>        
    <div class="navbar">
	
      <!-- Navigation item -->
      <ul>
        <li><a href="index.html">Home</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="overview.html">About</a></li>
      </ul>            
      <!-- Navigation item -->
      <ul>
        <li><a href="#">Download<!--[if IE 7]><!--></a><!--<![endif]-->
          <!--[if lte IE 6]><table><tr><td><![endif]-->
            <ul>
              <li><a href=downloads/express-1.2.1/express-1.2.1-macosx_x86_64.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Mac', 'FAQ']);" target="_blank">Mac OS X (64-bit)</a></li>
              <li><a href=downloads/express-1.2.1/express-1.2.1-linux_x86_64.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Linux', 'FAQ']);" target="_blank">Linux (64-bit)</a></li>
              <li><a href=downloads/express-1.2.1/express-1.2.1-win32_x86_64.zip onClick="_gaq.push(['_trackEvent', 'Downloads', 'Windows', 'FAQ']);" target="_blank">Windows (64-bit)</a></li>
              <li><a href=downloads/express-1.2.1/express-1.2.1-src.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Source', 'FAQ']);" target="_blank">Source Code</a></li>
              <li><a href="downloads" onClick="_gaq.push(['_trackEvent', 'Downloads', 'Previous','FAQ']);">Previous Versions</a></li>
           </ul>
          <!--[if lte IE 6]></td></tr></table></a><![endif]-->
        </li>
      </ul>                  
       <!-- Navigation item -->
      <ul>
        <li><a href="tutorial.html">Getting Started</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="manual.html">Manual</a></li>
      </ul>
      <ul>
        <li><a href="faq.html">FAQ</a></li>
      </ul>
      </div>
  
    <!-- C. MAIN SECTION -->      
    <div class="main">
    	<h1 class="pagetitle">FAQ</h1>

      <!-- C.1 CONTENT -->
      <div class="content">
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
          <h2 id="diff">Can eXpress be used for finding differentially expressed genes in RNA-Seq analysis?</h2>	
		  <p>eXpress does not test for differential expression but the (rounded) "effective counts" in the results.xprs file can be used in downstream count-based analysis tools such as <a href="http://www.bioconductor.org/packages/2.6/bioc/html/DEGseq.html">DEGSeq</a>, <a href="http://www.bioconductor.org/packages/release/bioc/html/edgeR.html">edgeR</a>, <a href="http://www.bioconductor.org/packages/release/bioc/html/baySeq.html">baySeq</a>, or <a href="http://www.bioconductor.org/packages/release/bioc/html/DESeq.html">DESeq</a>. However, none of these tools are suitable for isoform level differential expression and until our own differential expression tool is completed, we currently recommend <a href=http://cufflinks.cbcb.umd.edu>Cuffdiff</a> for that purpose.</p>
		  <p>&rarr; <a href="#top">Back to top.</a></p>
	<h2 id="assmbl">Can eXpress be used to perform genome or transcriptome assembly?</h2>	
	<p>No, eXpress is solely a <i>quantification</i> tool and cannot be used for assembly. However we believe the streaming algorithms underlying eXpress can be adapted for use in assembly as well and we are currently exploring ideas for low-memory streaming based assembly. For now we recommend <a href=http://cufflinks.cbcb.umd.edu>Cufflinks</a> for reference-based transcriptome assembly. <a href=http://trinityrnaseq.sourceforge.net/>Trinity</a>, <a href=http://www.ebi.ac.uk/~zerbino/oases/>Oases</a>, and <a href=http://www.bcgsc.ca/platform/bioinfo/software/trans-abyss>Trans-ABySS</a> can be used for <i>de novo</i> transcriptome assembly. There are numerous new short-read assembly tools that can be used for DNA assembly.</p>
	<p>&rarr; <a href="#top">Back to top.</a></p>
	<h2 id="restr">Is eXpress restricted to RNA-Seq analysis?</h2>
	<p>The methods in eXpress are general and can be used in any application where it is necessary to quantify the relative abundances of sequences that have been sampled using short reads. RNA-Seq is the application that we have focused on in initial testing because of the large amount of available data, however eXpress can be easily applied to allele-specific expression, metagenomics, and many other active areas of research.</p>
    <p>&rarr; <a href="#top">Back to top.</a></p>
	<h2 id="stream">Does eXpress only work in streaming mode?</h2>	
	<p>No, eXpress is also able to accept a SAM or BAM file as input (see <a href=manual.html#input>Manual</a> and <a href=tutorial.html#reqin>Getting Started</a> for more information on how to use different inputs. If given an alignment file, eXpress is also able to do multiple passes over the data to improve accuracy.  See the <a href=manual.html#running>Manual</a> for more details on different iterative modes.</p>
	<p>&rarr; <a href="#top">Back to top.</a></p>
   	<h2 id="single">Does eXpress accept both single-end and paired-end read alignments?</h2>
	<p>Yes, eXpress accepts both types of reads, even in the same alignment file.  However, to maximize accuracy of the bias correction and fragment length calculations, you should not mix reads generated in separate library preps. See <a href="#seppreps">below</a> for more details.</p>
    <p>&rarr; <a href="#top">Back to top.</a></p>

   	<h2 id="seppreps">Can I give eXpress a combined set of reads generated in different library preparations or with different read lengths?</h2>
	<p>Yes, but you have to do so carefully.  eXpress can take a comma-separated list of SAM filenames as its command-line input.  When this is done, eXpress will assume that each alignment file contains reads fir the same sample but sequenced from a different library preparation and/or using a different read length.  It will then estimate auxiliary parameters independently for each set but will combine the count/abundance results. Since some overall power is lost when smaller alignment sets are used, any reads generated from the same library preparation using the same read lengths should be combined into a single alignment file.  See the <a href="manual.html#sam">Manual</a> for more details.</p>
    <p>&rarr; <a href="#top">Back to top.</a></p>

   	<h2 id="strip">Why does eXpress say it processed 0 fragments when I gave it a bunch of paired-end reads?</h2>
	<p>This is most likely because the paired-end reads do not have matching names in your alignment file. eXpress knows that your reads are supposed to be paired-end in this case, but it assumes one end is unmapped and tosses the fragment when the matching reads cannot be found.  See the <a href="manual.html#sam">Manual</a> for details on how to address this issue.</p>
    <p>&rarr; <a href="#top">Back to top.</a></p>

    <h2 id="pub">When will the eXpress method be published?</h2>
	<p>We have submitted a manuscript and expect it to be published in the coming months.</p>
	<p>&rarr; <a href="#top">Back to top.</a></p>
    <h2 id="help">What do I do if I need help and my question isn't answered here?</h2>
	<p>First, you should check to see if you question is answered in the <a href="manual.html">Manual</a> or <a href="tutorial.html">Tutorial</a>.  If not, you should see if it has been addressed at the <a href=http://seqanswers.com>SeqAnswers</a> forum.  If you still can't find your answer, feel free to contact us at <a mailto:ask.xprs@gmail.com>ask.xprs@gmail.com</a>.  We are busy but will do our best to respond as soon as possible!</p>
    <p>&rarr; <a href="#top">Back to top.</a></p>
     </div> 
        <div class="corner-content-1col-bottom"></div>                               
      </div>
    
           <!-- C.2 SUBCONTENT -->
      <div class="subcontent">
      
          <!-- SUBCONTENT CELL -->
        <div class="corner-subcontent-top"></div>                        
        <div class="subcontent-box">
          <h1 class="menu">Questions</h1> 
          <div class="sidemenu1">
          <!-- CONTENT CELL -->                
 		  <ul>
   		  	<li><a href="#diff">Can eXpress be used for finding differentially expressed genes in RNA-Seq analysis?</a></li>
 		  	<li><a href="#assmbl">Can eXpress be used to perform genome or transcriptome assembly?</a></li>
			<li><a href="#restr">Is eXpress restricted to RNA-Seq analysis?</a></li>
			<li><a href="#stream">Does eXpress only work in streaming mode?</a></li>
			<li><a href="#single">Does eXpress accept both single-end and paired-end read alignments?</a></li>
			<li><a href="#seppreps">Can I give eXpress a combined set of reads generated in different library preparations or with different read lengths?</a></li>
			<li><a href="#strip">Why does eXpress say it processed 0 fragments when I gave it a bunch of paired-end reads?</a></li>
			<li><a href="#pub">When will the eXpress method be published?</a></li>
			<li><a href="#help">What do I do if I need help and my question isn't answered here?</a></li>
		  </ul>
	  </div> 
         </div>
        <div class="corner-subcontent-bottom"></div> 

 	 </div>
    </div>
      
    <!-- D. FOOTER -->      
    <div class="footer">
      <p>Copyright &copy; 2011 Adam Roberts&nbsp;&nbsp;|&nbsp;&nbsp;All Rights Reserved</p>
      <p class="credits">Design by <a href="http://1234.info/" title="Designer Homepage">1234.info</a> | Modified by <a href="http://cs.berkeley.edu/~adarob/">Adam Roberts</a> | <a href="http://validator.w3.org/check?uri=referer" title="Validate XHTML code">XHTML 1.0</a> | <a href="http://jigsaw.w3.org/css-validator/" title="Validate CSS code">CSS 2.0</a></p>
      <br />
      <p>The eXpress project was funded in part by an NSF graduate fellowship to Adam Roberts and NIH grant 1R01HG006129-01</p>
    </div>
    <div class="corner-page-bottom"></div>        
 </div> 
  
</body>
</html>



