<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en">

<!--  Version: Multiflex-5.4 / About                     -->
<!--  Type:    Design with sidebar                          -->
<!--  Date:    March 13, 2008                               -->
<!--  Design:  www.1234.info                                -->
<!--  License: Fully open source without restrictions.      -->
<!--           Please keep footer credits with the words    -->
<!--           "Design by 1234.info". Thank you!            -->

<head> 
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-25486066-1']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>
  <meta http-equiv="content-type" content="text/html; charset=utf-8" />
  <meta http-equiv="cache-control" content="no-cache" />
  <meta http-equiv="expires" content="3600" />
  <meta name="revisit-after" content="2 days" />
  <meta name="robots" content="index,follow" />
  <meta name="publisher" content="Your publisher infos here" />
  <meta name="copyright" content="Copyright (c) 2011 Adam Roberts" />
  <meta name="author" content="Designed by www.1234.info / Modified: Adam Roberts" />
  <meta name="distribution" content="global" />
  <meta name="image" content="http://bio.math.berkeley.edu/eXpress/img/logo.png" />
  <meta name="description" content="eXpress is a general quantification tool for target DNA/RNA sequences. While its primary use currently is RNA-Seq it has the potential for applications in many other areas including as allele-specific expression and metgenomics. What makes eXpress different is that it is an online (or streaming) algorithm, meaning it only makes one pass through the data. This allows it to be very light-weight and efficient using a constant amount of memory and time linear in the number of sequenced fragments being processed. Furthermore, it accepts piped SAM/BAM input, allowing users to avoid storing extremely large alignment files. eXpress models fragment biases (sequence-specific and positional), fragment lengths, and errors, allowing it to also be one of the most accurate quantification methods available." />
  <meta name="keywords" content="RNA-Seq, Genomics, transcript,quantification" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_reset.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_grid.css" />
  <link rel="stylesheet" type="text/css" media="screen,projection,print" href="./css/mf54_content.css" />
  <link rel="icon" type="image/x-icon" href="./img/favicon.ico" />
  <title>eXpress &bull; About</title>
   <script language="Javascript">
        logo1=new Image
        logo1.src="img/logo.png"
        logo2=new Image
        logo2.src="img/logo_white.png"
 </script>
  <title>eXpress &bull; About</title>
</head>

<!-- Global IE fix to avoid layout crash when single word size wider than column width -->
<!-- Following line MUST remain as a comment to have the proper effect -->
<!--[if IE]><style type="text/css"> body {word-wrap: break-word;}</style><![endif]-->



<body>
  <!-- CONTAINER FOR ENTIRE PAGE -->
  <div class="container">

    <!-- A. HEADER -->         
    <div class="corner-page-top"></div>        
    <div class="header">
      <div class="header-top">
        
        <!-- A.1 SITENAME -->    
        <div class="sitelogo">
        <ul>
            <li><a href="#" onMouseOver="document.logo.src=logo2.src" onMouseOut="document.logo.src=logo1.src"><img name="logo" src="img/logo.png"/></a></li>
        </ul>
        </div>
        <div class="sitename">
          <h1><a href="#">eXpress</a></h1> 
          &nbsp;
          <h2><i>Streaming</i> RNA-Seq Analysis</h2>
        </div>
    
        <!-- A.2 BUTTON NAVIGATION -->
        <div class="navbutton">
          <ul>
            <li><a href="http://www.berkeley.edu"><img src="img/berkeley_seal.gif"/></a></li>
          </ul>
        </div>
      </div>

      <!-- A.4 BREADCRUMB and SEARCHFORM -->
      <div class="header-bottom">

		

        <!-- Search form -->    
<div class="searchform" id="cse-search-form" style="padding-top:4px; width:30%;">Loading</div>
<script src="http://www.google.com/jsapi" type="text/javascript"></script>
<script type="text/javascript"> 
  google.load('search', '1', {language : 'en'});
  google.setOnLoadCallback(function() {
    var customSearchControl = new google.search.CustomSearchControl('017113487431020499685:jdftct1emrw');
    customSearchControl.setResultSetSize(google.search.Search.FILTERED_CSE_RESULTSET);
    var options = new google.search.DrawOptions();
    options.enableSearchboxOnly("http://www.google.com/cse?cx=017113487431020499685:jdftct1emrw", null, true);
    customSearchControl.draw('cse-search-form', options);
  }, true);
</script>
<link rel="stylesheet" href="css/googlesearch.css" type="text/css" />

   </div>
    </div>      
    <div class="corner-page-bottom"></div>    
    
    <!-- B. NAVIGATION BAR -->
    <div class="corner-page-top"></div>        
    <div class="navbar">
	
      <!-- Navigation item -->
      <ul>
        <li><a href="index.html">Home</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="overview.html">About</a></li>
      </ul>            
      <!-- Navigation item -->
      <ul>
        <li><a href="#">Download<!--[if IE 7]><!--></a><!--<![endif]-->
          <!--[if lte IE 6]><table><tr><td><![endif]-->
            <ul>
              <li><a href=downloads/express-0.9.1/express-0.9.1-src.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Source', 'About']);" target="_blank">Source</a></li>
              <li><a href=downloads/express-0.9.1/express-0.9.1-macosx_x86_64.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Mac', 'About']);" target="_blank">Mac OS X (64-bit)</a></li>
              <li><a href=downloads/express-0.9.1/express-0.9.1-linux_x86_64.tgz onClick="_gaq.push(['_trackEvent', 'Downloads', 'Linux', 'About']);" target="_blank">Linux (64-bit)</a></li>
              <li><a href="downloads" onClick="_gaq.push(['_trackEvent', 'Downloads', 'Previous','About']);">Previous Versions</a></li>
            </ul>
          <!--[if lte IE 6]></td></tr></table></a><![endif]-->
        </li>
      </ul>                  
       <!-- Navigation item -->
      <ul>
        <li><a href="tutorial.html">Getting Started</a></li>
      </ul>
      <!-- Navigation item -->
      <ul>
        <li><a href="manual.html">Manual</a></li>
      </ul>
      <ul>
        <li><a href="faq.html">FAQ</a></li>
      </ul>
      </div>
  
    <!-- C. MAIN SECTION -->      
    <div class="main">
    	<h1 class="pagetitle">About</h1>

      <!-- C.1 CONTENT -->
      <div class="content">

        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
        <h1 id="description">Description</h1>
			<p>eXpress is a streaming tool for quantifying the abundances of a set of target sequences from sampled subsequences. An example of an application is transcript-level
        RNA-Seq quantification. It is based on an online-EM
        algorithm [<a href="#ref1">1</a>] that results in space (memory) requirements
        proportional to the total size of the target sequences  and time
        requirements that are proportional to the number of sampled
        fragments. Thus, in applications such as RNA-Seq, eXpress can accurately quantify much larger
        samples than other currently available
        tools greatly reducing computing infrastructure requirements. eXpress can be used to build lightweight
			high-throughput sequencing  processing pipelines when coupled with
			a streaming aligner (such as Bowtie), as
			output can be piped directly into eXpress, effectively eliminating the need to store read alignments in either memory or disk.</p>

	  <p>In an analysis of the performance of eXpress for RNA-Seq analysis, we have observed that this efficiency does not come
			at a cost of accuracy. eXpress is more accurate
			than other available tools, even when limited
			to smaller datasets that do not require such
			efficiency [<a href="#ref2">2</a>]. Moreover, like the <a href=http://cufflinks.cbcb.umd.edu>Cufflinks</a>
			program [<a href="#ref3">3</a>], eXpress can be used to
			estimate transcript abundances in
			multi-isoform genes. eXpress is also able
			to resolve multi-mappings of reads across gene
			families, and does not require a reference
			genome so that it can be used in conjunction
			with <i>de novo</i> assemblers such as <a href=http://trinityrnaseq.sourceforge.net/>Trinity</a>, <a
        href=http://www.ebi.ac.uk/~zerbino/oases/>Oases</a>, or <a
        href=http://www.bcgsc.ca/platform/bioinfo/software/trans-abyss>Trans-ABySS</a>. The underlying model is based on previously described probabilistic
			models developed for RNA-Seq [<a href="#ref4">4</a>] but is applicable to other settings where target sequences are sampled, and includes
			parameters for fragment length distributions,
			errors in reads, and fragment bias [<a href="#ref5">5</a>].
	  </p>

			<p>eXpress can be used as to resolve ambiguous mappings in
			other high-throughput sequencing based applications. The only required
			inputs to eXpress are a set of target
			sequences and a set of sequenced fragments
			multiply-aligned to them.  While these
			target sequences will often be gene
			isoforms, they need not be.  Haplotypes could
			be used as the reference for allele-specific
			expression analysis or target genomes may be
			used in a metagenomics experiments.  eXpress will be useful in any analysis where reads multi-map to sequences that differ in abundance.</p>
      			<p>&rarr; <a href="#top">Back to top.</a></p>
      	</div>      

        <div class="corner-content-1col-bottom"></div>                               
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
		  	<h1 id="features">Features</h1>
		  	<ul>
		  	<li>Time proportional to number of reads, memory proportional to transcriptome size.</li>
		  	<li>Maximum of 3 free processor cores required (2 without bias
		  	correction). Try it on your laptop!</li>
		  	<li>Outputs <a href="http://cufflinks.cbcb.umd.edu/howitworks.html#hqua">FPKM</a>, estimated counts, and count
		  	variance/covariance for differential
		  	abundance analysis.</li>
		  	<li>Supports alignments with unlimited multi-mappings and errors.</li>
		  	<li>Transcript-level abundance estimation when used for RNA-Seq.</li>
			<li>Can be used for allele-specific expression estimates in RNA-Seq.</li>
		    <li>Corrects for both positional and sequence-specific fragment biases.</li>
   		    <li>Learns first-order Markov model for sequencing errors for use in probablistic read assignment.</li>
   		    <li>Learns fragment length distribution for use in probablistic assignment of paired-end reads.</li>
   		    <li>Supports both directional and non-directional sequencing.</li>
		  	<li>Written in C++ and runs on Mac, Linux, and Windows.</li>
		  	</ul>
      			<p>&rarr; <a href="#top">Back to top.</a></p>
      	</div>      

        <div class="corner-content-1col-bottom"></div>                               
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">

			<h1 id="methods">Methods</h1>
			<p>A complete description of the eXpress method will be found in the forthcoming manuscript [<a href="ref1">2</a>].
			      		<p>&rarr; <a href="#top">Back to top.</a></p>	
			      </div>      

        <div class="corner-content-1col-bottom"></div>                               
        <!-- CONTENT CELL -->                
        <div class="corner-content-1col-top"></div>                        
        <div class="content-1col-nobox">
			<h1 id="references">References</h1>
	  		<ol>
	  	  <li id="ref1">Cappé O and Moulines E. (2009).
	  	  <b>On-line expectation–maximization algorithm for latent data models.</b>
	  	  <i>Journal of the Royal Statistical Society.</i>
	  	  <br />&rarr;<a href=http://http://onlinelibrary.wiley.com/doi/10.1111/j.1467-9868.2009.00698.x/abstract>doi:10.1111/j.1467-9868.2009.00698.x</a></li>
	  		<li id="ref2">Roberts A and Pachter L (2011).
	  		<b>Efficient and accurate transcript-level abundance estimation from RNA-Seq reads using the online EM algorithm.</b>
	  		<i>In preparation.</i></li>
	  	  <li id="ref3">Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B,&nbsp; Pachter L (2010).
	  	  <b>Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation.</b>
	  	  <i>Nature Biotechnology.</i>
	  	  <br />&rarr;<a href=http://dx.doi.org/10.1038/nbt.1621>doi:10.1038/nbt.162</a></li>
	  	  <li id="ref4">Pachter, L (2011).
	  	  <b>Models for transcript quantification from RNA-Seq.</b>
	  	  <i>Submitted.</i>
          <br />&rarr;<a href="http://arxiv.org/abs/1104.3889">arXiv:1104.3889v2</a></font></li>
	  	  </li>
         <li id="ref5">Roberts A, Trapnell C, Donaghey J, Rinn JL, Pachter L (2011).
          <b>Improving RNA-Seq expression estimates by correcting for fragment bias.</b>
          <i>Genome Biology.</i>
          <br />&rarr;<a href="http://genomebiology.com/2011/12/3/R22/abstract">doi:10.1186/gb-2011-12-3-r22</a></font></li>
      	</ol>
      		<p>&rarr; <a href="#top">Back to top.</a></p>
      </div>      

        <div class="corner-content-1col-bottom"></div>                               
      </div>
                
       <!-- C.2 SUBCONTENT -->
      <div class="subcontent">
      
         <!-- SUBCONTENT CELL -->
        <div class="corner-subcontent-top"></div>                        
        <div class="subcontent-box">
          <h1 class="menu">Outline</h1> 
          <div class="sidemenu1">
          <!-- CONTENT CELL -->                
 		  <ul>
			  <li><a href="#description">Description</a></li>
			  <li><a href="#features">Features</a></li>
			  <li><a href="#method">Method</a></li>
			  <li><a href="#references">References</a></li>

		  </ul>
          </div> 
          </div>
        <div class="corner-subcontent-bottom"></div> 
	</div>
 	 </div>
      
    <!-- D. FOOTER -->      
    <div class="footer">
      <p>Copyright &copy; 2011 Adam Roberts&nbsp;&nbsp;|&nbsp;&nbsp;All Rights Reserved</p>
      <p class="credits">Design by <a href="http://1234.info/" title="Designer Homepage">1234.info</a> | Modified by <a href="http://cs.berkeley.edu/~adarob/">Adam Roberts</a> | <a href="http://validator.w3.org/check?uri=referer" title="Validate XHTML code">XHTML 1.0</a> | <a href="http://jigsaw.w3.org/css-validator/" title="Validate CSS code">CSS 2.0</a></p>
      <br />
      <p>The eXpress project was funded in part by an NSF graduate fellowship to Adam Roberts and NIH grant 1R01HG006129-01</p>
    </div>
    <div class="corner-page-bottom"></div>        
  </div> 
  
</body>
</html>



